src/str.js

```diff
@@ -716,6 +755,162 @@ Sk.builtin.str.prototype["isdigit"] = new Sk.builtin.func(function (self) {
+            if(Math.max([expanded.lastIndexOf("\r"),expanded.lastIndexOf("\n")]) !== -1){
+                spacerem = tabsize - ((expanded.length - (Math.max(expanded.lastIndexOf("\r"),expanded.lastIndexOf("\n")))-1) % tabsize);
+            } else {
+                spacerem = tabsize - (expanded.length % tabsize);
+            }
+        }
+        spacestr = goog.string.repeat(" ", spacerem);
+        if(x!==split.length-1){
+            expanded += split[x] +  spacestr;
+        } else{
+            expanded += split[x];
+        }
+
+    }
+
+    return new Sk.builtin.str(expanded);
+});
+
+Sk.builtin.str.prototype["swapcase"] = new Sk.builtin.func(function (self) {
+    var i;
+    var letters;
+    var newletters = "";
+    var ch;
+    Sk.builtin.pyCheckArgs("swapcase", arguments, 1, 1);
+    letters = self.v;
+
+    if (letters.length === 0){
+        return new Sk.builtin.str("");
+    }
+
+    for(i = 0; i<letters.length; i++){
+        ch = letters.charAt(i);
+        if(letters[i] === ch.toLowerCase()){
+            newletters += ch.toUpperCase();
+        }else {
+            newletters += ch.toLowerCase();
+        }
+    }
+    return new Sk.builtin.str(newletters);
+});
+
+Sk.builtin.str.prototype["splitlines"] = new Sk.builtin.func(function (self, keepends) {
+    var data = self.v;
+    var i = 0;
+    var j = i;
+    var selflen = self.v.length;
+    var strs_w = [];
+    var ch;
+    var eol;
+    var sol = 0;
+    var slice;
+    Sk.builtin.pyCheckArgs("splitlines", arguments, 1, 2);
+    if ((keepends !== undefined) && !Sk.builtin.checkBool(keepends)) {
+        throw new Sk.builtin.TypeError("boolean argument expected, got " + Sk.abstr.typeName(keepends));
+    }
+    if (keepends === undefined){
+        keepends = false;
+    } else {
+    keepends = keepends.v;
+    }
+
+
+    for(i=0; i<selflen; i++){
+        ch = data.charAt(i);
+        if(data.charAt(i+1) ==="\n" && ch === "\r"){
+            eol = i +2;
+            slice = data.slice(sol, eol);
+            if(!keepends){
+                slice = slice.replace(/(\r|\n)/g,"");
+            }
+            strs_w.push(new Sk.builtin.str(slice));
+            sol = eol;
+            
+        }else if((ch === "\n" && data.charAt(i-1) !== "\r") || ch === "\r"){
+            eol = i + 1;
+            slice = data.slice(sol, eol);
+            if(!keepends){
+                slice = slice.replace(/(\r|\n)/g,"");
+            }
+            strs_w.push(new Sk.builtin.str(slice));
+            sol = eol;
+        }
+
+    }
+    if(sol < selflen){
+        eol = selflen;
+        slice = data.slice(sol, eol);
+        if(!keepends){
+            slice = slice.replace(/(\r|\n)/g,"");
+        }
+        strs_w.push(new Sk.builtin.str(slice));
+    }
+
+    return new Sk.builtin.list(strs_w);
+});
+
+Sk.builtin.str.prototype["title"] = new Sk.builtin.func(function (self) {
+    var input = self.v;
+    var buffer = new Array(input.length).join(" ").split("");
```

Comment:
Populating an empty array without using map and an anonymous function. It's also quite a bit faster with very large strings.